runtime/near-vm-runner/src/logic/logic.rs

```diff
@@ -30,14 +30,178 @@ fn base64(s: &[u8]) -> String {
     base64::engine::general_purpose::STANDARD.encode(s)
 }
 
+/// Structure representing the results and outcomes of a contract execution.
+///
+/// This is a subset of [`VMLogic`] that's strictly necessary to produce `VMOutcome`s.
+pub struct ExecutionResultState {
+    /// All gas and economic parameters required during contract execution.
+    config: Arc<Config>,
+    /// Gas tracking for the current contract execution.
+    gas_counter: GasCounter,
+    /// Logs written by the runtime.
+    logs: Vec<String>,
+    /// Tracks the total log length. The sum of length of all logs.
+    total_log_length: u64,
+    /// What method returns.
+    return_data: ReturnData,
+    /// Keeping track of the current account balance, which can decrease when we create promises
+    /// and attach balance to them.
+    current_account_balance: Balance,
+    /// Storage usage of the current account at the moment
+    current_storage_usage: StorageUsage,
+}
+
+impl ExecutionResultState {
+    pub fn new(context: &VMContext, config: Arc<Config>) -> Self {
+        // Overflow should be checked before calling VMLogic.
```

Comment:
Nit: add a doc-comment to `new` and `debug_assert!` it just to be sure. That being said, in practice I don’t think it could be a problem, as overflowing the amount would likely require more than the total supply of $NEAR (I hope)

Reply:
There's an easy fix, and something that we've agreed upon a long time ago: `checked_add` and `panic`. I don't trust "should be" comments inside the function implementation, and `VMLogic` (or `ExecutionResultState`s) ain't being created by a bucketloads in a hot >1000Hz loops anyway...